File: src/test/java/de/dk/api/codeChallenge/notification/service/NotificationTypeServiceTest.java
package de.dk.api.codeChallenge.notification.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import de.dk.api.codeChallenge.notification.model.NotificationTypeEntity;
import de.dk.api.codeChallenge.notification.model.enums.Category;
import de.dk.api.codeChallenge.notification.repository.NotificationTypeRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class NotificationTypeServiceTest {

    private NotificationTypeRepository repository;
    private NotificationTypeService service;

    @BeforeEach
    void setup() {
        repository = mock(NotificationTypeRepository.class);
        service = NotificationTypeService(repository);
    }

    private NotificationTypeService NotificationTypeService(NotificationTypeRepository repository) {
        return new NotificationTypeService(repository);
    }

    @Test
    void createOrGet_givenExistingType_shouldReturnExisting() {
        NotificationTypeEntity existing = new NotificationTypeEntity("type1", Category.A);
        when(repository.findByNameIgnoreCase("type1")).thenReturn(existing);

        NotificationTypeEntity result = service.createOrGet("type1", "A");

        assertEquals(existing, result);
        verify(repository, never()).save(any());
    }

    @Test
    void createOrGet_givenNewType_shouldSaveAndReturn() {
        when(repository.findByNameIgnoreCase("type2")).thenReturn(null);
        when(repository.save(any())).thenAnswer(invocation -> invocation.getArgument(0));

        NotificationTypeEntity result = service.createOrGet("type2", "B");

        assertEquals("type2", result.getName());
        assertEquals(Category.B, result.getCategory());
        verify(repository).save(any());
    }
}

File: src/test/kotlin/de/dkb/api/codeChallenge/notification/service/NotificationTypeServiceTest.kt
package de.dkb.api.codeChallenge.notification.service

import de.dkb.api.codeChallenge.notification.model.enums.Category
import de.dkb.api.codeChallenge.notification.model.NotificationTypeEntity
import de.dkb.api.codeChallenge.notification.repository.NotificationTypeRepository
import io.mockk.*
import org.junit.Assert.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class NotificationTypeServiceTest {

    private val repository = mockk<NotificationTypeRepository>(relaxed = true)
    private lateinit var service: NotificationTypeService

    @BeforeEach
    fun setup() {
        service = NotificationTypeService(repository)
    }

    @Test
    fun createOrGet_givenExistingType_shouldReturnExisting() {
        val existing = NotificationTypeEntity(name = "type1", category = Category.A)
        every { repository.findByNameIgnoreCase("type1") } returns existing

        val result = service.createOrGet("type1", "A")

        assertEquals(existing, result)
        verify(exactly = 0) { repository.save(any()) }
    }

    @Test
    fun createOrGet_givenNewType_shouldSaveAndReturn() {
        every { repository.findByNameIgnoreCase("type2") } returns null
        every { repository.save(any()) } answers { firstArg() }

        val result = service.createOrGet("type2", "B")

        assertEquals("type2", result.name)
        assertEquals(Category.B, result.category)
        verify { repository.save(any()) }
    }
}
